src/sentry/services/hybrid_cloud/access/service.py

```diff
@@ -0,0 +1,138 @@
+            return False
+        return True
+
+    def query_sso_state(
+        self,
+        *,
+        organization_id: Optional[int],
+        is_super_user: bool,
+        member: Optional[RpcOrganizationMemberSummary],
+    ) -> RpcMemberSsoState:
+        if organization_id is None:
+            return _SSO_NONMEMBER
+
+        # we special case superuser so that if they're a member of the org they must still follow SSO checks
+        # or put another way, superusers who are not members of orgs bypass SSO.
+        if member is None or member.user_id is None:
+            if is_super_user:
+                return _SSO_BYPASS
+            return _SSO_NONMEMBER
+
+        auth_provider = self.get_auth_provider(organization_id=organization_id)
+        if auth_provider is None:
+            return _SSO_BYPASS
+
+        if auth_provider.flags.allow_unlinked:
+            return _SSO_BYPASS
+        else:
+            requires_sso = True
+
+            auth_identity = self.get_auth_identity_for_user(
+                auth_provider_id=auth_provider.id, user_id=member.user_id
+            )
+            if auth_identity is None:
+                sso_is_valid = False
+                requires_sso = not self.can_override_sso_as_owner(auth_provider, member)
+            else:
+                sso_is_valid = self.auth_identity_is_valid(
+                    auth_identity=auth_identity, member=member
+                )
+
+        return RpcMemberSsoState(is_required=requires_sso, is_valid=sso_is_valid)
+
+    def get_user_auth_state(
+        self,
+        *,
+        user_id: int,
+        is_superuser: bool,
+        organization_id: Optional[int],
+        org_member: Optional[RpcOrganizationMemberSummary],
+    ) -> RpcAuthState:
+        sso_state = self.query_sso_state(
+            organization_id=organization_id, is_super_user=is_superuser, member=org_member
+        )
+
+        if is_superuser:
+            # "permissions" is a bit of a misnomer -- these are all admin level permissions, and the intent is that if you
+            # have them, you can only use them when you are acting, as a superuser.  This is intentional.
+            permissions = list(self.get_permissions_for_user(user_id))
+        else:
+            permissions = []
+
+        return RpcAuthState(sso_state=sso_state, permissions=permissions)
+
+    @abc.abstractmethod
+    def get_permissions_for_user(self, user_id: int) -> FrozenSet[str]:
+        pass
+
+
+def impl_by_region_resources() -> AccessService:
+    from sentry.services.hybrid_cloud.access.impl import RegionAccessService
+
+    return RegionAccessService()
+
+
+def impl_by_control_resources() -> AccessService:
+    from sentry.services.hybrid_cloud.access.impl import ControlAccessService
+
+    return ControlAccessService()
+
+
+access_service: AccessService = silo_mode_delegation(
+    {
+        SiloMode.REGION: impl_by_region_resources,
+        SiloMode.CONTROL: impl_by_control_resources,
```

Comment:
Do we have any way to validate that the two implementations behave the same given a consistent starting point with replicated records?

Reply:
Yes, in the sense that test_access already does this by using the same suite for all modes.